### DISCUSSION TITLE
Are there any plans to do anything specific for console apps with Project Reunion?

### DISCUSSION THREAD

**Discussion by mrlacey** [microsoft/WindowsAppSDK#874] [General] [answered]
Console applications are only mentioned in passing in relation to Project Reunion and WinUI3. Are there plans for them in relation to ProjectReunion?

It's not an option as a project type in the VSIX. Are there plans that it will ever be there?

Is this question WinUI(3) specific? - console apps were possible with UWP  but not very (AFAIK) widely used. Does this mean they'll come to WinUI3 too? Or is WinUI3 only focused on code with a **UI**?

Or do console apps fall under .NET as a project type with no plans or need for anything Windows specific?

**Comment by mveril** [comment]
If I understand your question correctly, I think it is possible to use Reunion API for any kind of applications. I think you can consume it (or a part of it) as a [nuget package](https://www.nuget.org/packages/Microsoft.ProjectReunion) A good example is the push notification test console app used at the "[What's new for Windows desktop application development](https://mybuild.microsoft.com/sessions/79b87707-0505-4144-aab9-94cc30f91340)" session of MSBuild 2021 but WinUI 3 is a UI framework and so for me there is no sense in using it on a console app.  if you want to use WinRT API you can already do it in [.NET 5 using the TFM](https://blogs.windows.com/windowsdeveloper/2020/09/03/calling-windows-apis-in-net5) But maybe I misunderstood your question.

**Reply by orcmid** [reply]
I thought the use of a console app as a way to explore and confirm some operations that would later be carried out in a (cloud-based) GUI app was great.  It's a terrific way to confirm some things at the model level before investing development on a full end-user-facing (enterprise) app.  Of course it even demonstrated continuing use of PInvoke.

I found this interesting with regard to some single-player text-adventure/found-object game development that can be established initially with a command-line app and then, thanks to the capabilities of Windows Terminal, maybe elevated to an (MS-DOS style) CUA ASCII-terminal app.  Going to a level that is untethered from the desktop is a different challenge so I don't know anything in that direction would qualify as being a console app (as opposed to also being MacOS and Linux desktop ported).

I'm also unclear what @mrlacey means by a console application that has much dependency on Project Reunion, although I can imagine some dependencies as one moves to cloud-based something-or-other.  At some point it is not a console app any longer.

**Reply by mrlacey** [reply]
My original intention was to leave the question open, rather than try and force the discussion in a single direction.
Console apps have long been part of Windows and with Project Reunion being the "next evolution" of components and tools for Windows app development I thought the omission of any reference, plan, or discussion relating to console apps to be notable.
Do they have an evolution or tooling consideration as part of Project Reunion?
Are they done?
Is this something that might be visited in the future? (if so, where in the backlog is it?)

WinUI 3 is part of Project Reunion and also the successor to UWP. UWP had the ability to produce console apps within the UWP app model. 
WinUI 3 is a UX framework (not just a UI framework.) It includes things like the app model which can have an impact on what's done with console apps.

If the intention and expectation is that Project Reunion APIs will be able to be consumed in a .NET 6 command line app  targeting a Windows-specific TFM then that will be fine to know.

I'm not expecting answers right now, just an official acknowledgement that there needs to be some clarity on how console apps relate to Project Reunion.

At the moment all I have is speculation and guesswork about what might work, what is expected to work, and what may one day come.
As is noted above there are lots of varied scenarios where console apps can be useful. It would be a shame to see thought and consideration for them to be lost.

**Reply by orcmid** [reply]
> Console apps have long been part of Windows and with Project Reunion being the "next evolution" of components and tools for Windows app development I thought the omission of any reference, plan, or discussion relating to console apps to be notable.
> ...
> At the moment all I have is speculation and guesswork about what might work, what is expected to work, and what may one day come.

@mrlacey Have you looked at the separate [Windows Console and Terminal Ecosystem Roadmap](https://docs.microsoft.com/en-us/windows/console/ecosystem-roadmap).   This might also help clarify what is and what is not a console application.  I always think command-line but maybe you mean something else?

**Reply by ghost** [reply]
@mrlacey

> If the intention and expectation is that Project Reunion APIs will be able to be consumed in a .NET 6 command line app targeting a Windows-specific TFM then that will be fine to know.

 [#419-Support PowerShell scripts using Reunion APIs](https://github.com/microsoft/ProjectReunion/issues/419)  
this one hardly tied to powershell only but no mention of plans regarding an evolutuion of generic console app model.  :(

**Reply by mdtauk** [reply]
There was a session which demonstrated Reunion push notifications triggering the launch of a console app, as well as background app support.

Maybe it would be worth adding to the discussion, some suggestions of things you'd like to see come to console, or current console APIs which are cumbersome or difficult to work with, that could do with simplification or empowering with better features?

**Reply by mveril** [reply]
> My original intention was to leave the question open, rather than try and force the discussion in a single direction.

Sorry if I gave the impression of being narrow-minded.
> Console apps have long been part of Windows and with Project Reunion being the "next evolution" of components and tools for Windows app development I thought the omission of any reference, plan, or discussion relating to console apps to be notable.

It is certain that some project reunion component like AppLifecycle or push notification can be interesting even in the case of a console application but maybe console specific APIs will also be part of Project Reunions why not.
 
> WinUI 3 is part of Project Reunion and also the successor to UWP. UWP had the ability to produce console apps within the UWP app model.
> WinUI 3 is a UX framework (not just a UI framework.) It includes things like the app model which can have an impact on what's done with console apps.

As far as WinUI is concerned, as its name indicates the UI it's ony the successor of UWP XAML so I don't see the usefulness that it can have for console applications, indeed we mustn't confuse UWP and WinUI (WinUI concern the UI/UX layer and UWP is an app model). If I understand correctly the ultimate goal of Project Reunion by combining all its components is to merge the two app models to what is currently called the Reunion application model but WinUI concern only the UI layer.  Interesting Issues for the future Reunion app model are #491 and #829 

> If the intention and expectation is that Project Reunion APIs will be able to be consumed in a .NET 6 command line app targeting a Windows-specific TFM then that will be fine to know.

I think we only need to use the nuget packages of the reunion components we want except if you want an WinUI app. In this case we need to use a WinUI project template.
> 
> I'm not expecting answers right now, just an official acknowledgement that there needs to be some clarity on how console apps relate to Project Reunion.
> 
> At the moment all I have is speculation and guesswork about what might work, what is expected to work, and what may one day come.
> As is noted above there are lots of varied scenarios where console apps can be useful. It would be a shame to see thought and consideration for them to be lost.

I totally agree that we need clear documentation on how to use separate Reunion component without starting with a WinUI project template, but this is just the beginning.
In addition, the evolution of the UWP app model with future UWP apps built on .NET 6 (or perhaps a more distant version of .NET) is not at all clear. I was hoping for a clarification on this at the MSBuild 2021 conference but I haven't seen anything like this.

**Reply by orcmid** [reply]
> @mrlacey
> but no mention of plans regarding an evolutuion of generic console app model. :(

@mrlacey @mveril @ecovio1 
@orcmid 
> Have you looked at the separate [Windows Console and Terminal Ecosystem Roadmap](https://docs.microsoft.com/en-us/windows/console/ecosystem-roadmap). This might also help clarify what is and what is not a console application. I always think command-line but maybe you mean something else?

Although from 2018, it appears that the deprecations and switch to virtual terminals are underway.  Also see [What is Windows Terminal](https://docs.microsoft.com/en-us/windows/terminal/) and the project [is in the open here](https://github.com/microsoft/terminal).

**Reply by mveril** [reply]
I don't know if the UWP console apps are depreciated or not, but if they are not, it might be interesting to do a wrapper around [setConsoleMode](https://docs.microsoft.com/en-us/windows/console/setconsolemode) as part of Project Reunion to allow UWP app to use [virtual terminals APIs](https://docs.microsoft.com/windows/console/console-virtual-terminal-sequences) (maybe it is already possible to use it but I don't believe it) and Win32 apps to enable it easily. (Now it is enabled by default since 20H2 microsoft/terminal#9432).

The [Project reunion sample repo](https://github.com/microsoft/Project-Reunion-Samples) has a sample project for a [console unpackaged app that uses MrtCore](https://github.com/microsoft/Project-Reunion-Samples/tree/main/MrtCore/console_unpackaged_app).

**Comment by andrewleader** [comment, marked as answer]
We're certainly thinking about console apps, and [at the Build session](https://mybuild.microsoft.com/sessions/79b87707-0505-4144-aab9-94cc30f91340) last week, we demo'd a console app using push notifications (at around time marker 16:25 or so).

We don't have any timelines or details to share yet about project templates, but this is definitely something that we're looking at. Any feedback about whether you'd need console templates, or how you think project templates could best be exposed, is welcomed!
